File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
/*
    keep suming the elements and if the current sum is greater than max then update the max.
    if sum < 0 then we can freely remove them as they are only being a dead weight to the maximum sum

*/
    
    int maxSubArray(vector<int>& nums) {
        int sum = nums[0],mx = nums[0];
        for(int i=1;i<nums.size();i++){
            if (sum<0) sum = 0;
            sum+= nums[i];
            if (sum>mx)mx = sum;
        }
        return mx;
    }
    
};

File: 53-maximum-subarray/53-maximum-subarray.py
class Solution:
    """
    keep suming the elements and if the current sum is greater than max then update the max.
    if sum < 0 then we can freely remove them as they are only being a dead weight to the maximum sum
    """
    def maxSubArray(self, nums: List[int]) -> int:
        sm = 0;mx = -sys.maxsize
        for i in nums:
            sm += i
            mx = max(mx,sm)
            if sm<0 : sm = 0
        return mx